tests/std/tests/VSO_0000000_regex_use/test.cpp

```diff
@@ -680,6 +680,338 @@ void test_gh_5160() {
+    g_regexTester.should_not_match("^*a*", R"-(.*\(^*a*\))-", basic_or_grep);
+
+    // Validate that there is no special behavior near bars,
+    // as they are alternation operators in regex modes other than basic or grep.
+    {
+        const test_regex middle_bar(&g_regexTester, "^a|a", basic_or_grep);
+        middle_bar.should_search_match("a|a", "a|a");
+        middle_bar.should_search_fail("^a|a");
+        middle_bar.should_search_fail("ba|a");
+        middle_bar.should_search_fail("a");
+    }
+
+    {
+        const test_regex group_middle_bar(&g_regexTester, "^\\(a|a\\)", basic_or_grep);
+        group_middle_bar.should_search_match("a|a", "a|a");
+        group_middle_bar.should_search_fail("^a|a");
+        group_middle_bar.should_search_fail("ba|a");
+        group_middle_bar.should_search_fail("a");
+    }
+
+    {
+        const test_regex middle_bar_with_caret(&g_regexTester, "^a|^b", basic_or_grep);
+        middle_bar_with_caret.should_search_match("a|^b", "a|^b");
+        middle_bar_with_caret.should_search_fail("a|b");
+        middle_bar_with_caret.should_search_fail("^a|^b");
+        middle_bar_with_caret.should_search_fail("ca|^b");
+        middle_bar_with_caret.should_search_fail("a");
+        middle_bar_with_caret.should_search_fail("b");
+    }
+
+    {
+        const test_regex group_middle_bar_with_caret(&g_regexTester, "^\\(a|^b\\)", basic_or_grep);
+        group_middle_bar_with_caret.should_search_match("a|^b", "a|^b");
+        group_middle_bar_with_caret.should_search_fail("a|b");
+        group_middle_bar_with_caret.should_search_fail("^a|^b");
+        group_middle_bar_with_caret.should_search_fail("ca|^b");
+        group_middle_bar_with_caret.should_search_fail("a");
+        group_middle_bar_with_caret.should_search_fail("b");
+    }
+
+    g_regexTester.should_match("ab", "a|*b", basic_or_grep);
+    g_regexTester.should_match("a||b", "a|*b", basic_or_grep);
+    g_regexTester.should_not_match("a|*b", "a|*b", basic_or_grep);
+    g_regexTester.should_throw("a|**b", error_badrepeat, basic_or_grep);
+
+    g_regexTester.should_match("ab", "^a|*b", basic_or_grep);
+    g_regexTester.should_match("a||b", "^a|*b", basic_or_grep);
+    g_regexTester.should_not_match("a|*b", "^a|*b", basic_or_grep);
+    g_regexTester.should_throw("^a|**b", error_badrepeat, basic_or_grep);
+
+    g_regexTester.should_match("a|b", "^a|^*b", basic_or_grep);
+    g_regexTester.should_match("a|^^b", "^a|^*b", basic_or_grep);
+    g_regexTester.should_not_match("a|*b", "^a|^*b", basic_or_grep);
+    g_regexTester.should_not_match("a|^*b", "^a|^*b", basic_or_grep);
+    g_regexTester.should_throw("^a|^**b", error_badrepeat, basic_or_grep);
+}
+
+void test_gh_5165() {
+    // GH-5165: Revise caret parsing in basic and grep mode
+    test_gh_5165_syntax_option(basic);
+    test_gh_5165_syntax_option(grep);
+
+    // test cases specific for basic regular expressions

```

Comment:
Suggestion: I believe it would be clearer to extract the basic-only and grep-only tests into separate functions, instead of different regions of one big function.